Add tests of SamplingDataset and fix a bug

frreiss opened the pull request:
This PR adds regression tests for the SamplingDataset kernel, following the example of the tests for AssertNextDataset. @feihugis helped guide me through this task. To allow my tests to create instances of the SamplingDataset class, I refactored the SamplingDataset class from a single .cc file into a header file and a code file. I also fixed a bug in the dataset's save/restore code that was preventing one of the tests from running.

I've broken the refactoring, the tests, and the bug fixes into separate commits.

frreiss committed:
Refactor SamplingDataset into .cc and .h files

Refactor SamplingDataset into .cc and .h files

Additional refactoring

frreiss committed:
Fix compiler warnings

frreiss committed:
Add tests of SamplingDataset

frreiss committed:
Fix bug in save/restore of SamplingDataset

frreiss committed:
Fix formatting with clang-format and buildifier

Run source files through clang-format

Run sampling_dataset_op.[cc|h] through clang-format

Run simple_philox.h through clang-format

Run test case through clang-format

Run build file through buildifier

frreiss committed:
Merge branch 'master' of https://github.com/tensorflow/tensorflow into issue-data-sample-test

frreiss committed:
Correct error in merge

frreiss committed:
Fix additional problems from merge

frreiss committed:
Add reference to location of API definition

frreiss committed:
Merge branch 'master' of https://github.com/tensorflow/tensorflow into issue-data-sample-test

jsimsa commented on tensorflow/core/kernels/data/experimental/sampling_dataset_op_test.cc:
```diff
@@ -0,0 +1,733 @@
+};
+
+// Common parameters that every test case in this file shares
+struct TestCase {
```
Please follow the style @feihugis and I converged on, having separate test case structs for different types of tests (https://github.com/tensorflow/tensorflow/pull/31105).

feihugis replied:
@jsimsa Thanks for merging #31105! It will be helpful for us to refactor other tests!

feihugis replied:
@frreiss More examples can be found [here](https://github.com/tensorflow/tensorflow/pull/31344).

frreiss replied:
Thanks for the example code @feihugis. I have pushed a refactored version of the test case into the new format in commit edf1be3

feihugis replied:
@frreiss Thanks for updating the test format! There may be some more changes for the format. Will let you know once it is finalized.

feihugis replied:
@frreiss The changes in test formats are finished. An example can be found [here](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/kernels/data/batch_dataset_op_test.cc). Please feel free to let me know if you have any question or need any help.

frreiss replied:
Thanks @feihugis. I've restructured the tests in this PR to match the new format and checked the changes into commit d827bfa.

jsimsa commented on tensorflow/core/kernels/data/experimental/sampling_dataset_op.cc:
```diff
@@ -23,202 +25,207 @@ limitations under the License.
-    const DatasetBase* const input_;
+    // random util
+    random::PhiloxRandom parent_generator_ GUARDED_BY(mu_);
+    random::SimplePhilox generator_ GUARDED_BY(mu_);
```
If you make this `random::SingleSampleAdapter<random::PhiloxRandom>` similar to what `shuffle_dataset_op.cc` does, you will not need to change the API of `random::SimplePhilox` which would be good to avoid.

frreiss replied:
I've pushed some new changes in commit f1406b4 that make `SampleDataset` use `random::SingleSampleAdapter<random::PhiloxRandom>`.

jsimsa reviewed (changes requested):
Thank you @frreiss, left a couple of comments.

frreiss committed:
Merge branch 'master' of https://github.com/tensorflow/tensorflow into issue-data-sample-test

frreiss committed:
Refactor test case to new format

Refactor test case to new format

Refactor initialization of input dataset

frreiss committed:
Switch from SimplePhilox to SingleSampleAdapter

jsimsa commented on tensorflow/core/kernels/data/experimental/sampling_dataset_op_test.cc:
```diff
@@ -0,0 +1,847 @@
+  Tensor input_dataset;
+};
+
+SamplingDatasetParams OneHundredPercentSampleDataset() {
```
Values that do not change across the different test cases (seed, seed2, start, step) should not be an argument. `stop` can be replaced with `num_elements`.

frreiss replied:
Thanks for that feedback. I've pushed commit 02d9d92, which replaces the `seed`, `seed2`, `start`, and `step` parameters with constants and renames `stop` to `num_elements`.

jsimsa commented on tensorflow/core/kernels/data/experimental/sampling_dataset_op_test.cc:
```diff
@@ -0,0 +1,847 @@
+          /*expected_outputs=*/{}};
+}
+
+TEST_P(ParameterizedGetNextSamplingDatasetOpTest, GetNext) {
```
I believe that @feihugis is working on utilities for factoring out the commonalities in test setup. Fei, would you refactoring be able to avoid the duplication in code in the tests here?

feihugis replied:
@jsimsa Yes, the refactoring I'm working on will avoid the duplication here. Once the solution is finalized, I will work with @frreiss to remove the duplications here.

frreiss replied:
I've removed the duplicate initialization code as part of the refactoring in commit d827bfa.

frreiss committed:
Replaced test parameters that don't change with constants and renamed 'stop' parameter to 'num_elements'

jsimsa reviewed (commented):
LGTM, modulo reflecting refactoring by @feihugis

frreiss committed:
Merge branch 'master' of https://github.com/tensorflow/tensorflow into issue-data-sample-test

frreiss committed:
Clean-up after merge

frreiss committed:
Refactor tests into new format

frreiss committed:
Additional cleanup after refactoring

Additional cleanup after refactoring

Additional cleanup

frreiss commented:
Sorry, @jsimsa, I think I may have removed some review comments from you while squashing commits. Did you request any changes?

mellanox-github commented:
Can one of the admins verify this patch?

jsimsa commented:
Fails internal tests with:

```
In module '//testing/base/public:gunit_headers':
third_party/googletest/googletest/include/gtest/internal/gtest-param-util.h:401:16: error: allocating an object of abstract class type 'tensorflow::data::experimental::(anonymous namespace)::ParameterizedGetNextTest_GetNext_Test'
    return new TestClass();
               ^
third_party/googletest/googletest/include/gtest/internal/gtest-param-util.h:397:12: note: in instantiation of member function 'testing::internal::ParameterizedTestFactory<tensorflow::data::experimental::(anonymous namespace)::ParameterizedGetNextTest_GetNext_Test>::CreateTest' requested here
  explicit ParameterizedTestFactory(ParamType parameter) :
           ^
third_party/googletest/googletest/include/gtest/internal/gtest-param-util.h:439:16: note: in instantiation of member function 'testing::internal::ParameterizedTestFactory<tensorflow::data::experimental::(anonymous namespace)::ParameterizedGetNextTest_GetNext_Test>::ParameterizedTestFactory' requested here
    return new ParameterizedTestFactory<TestSuite>(parameter);
               ^
third_party/googletest/googletest/include/gtest/internal/gtest-param-util.h:436:3: note: in instantiation of member function 'testing::internal::TestMetaFactory<tensorflow::data::experimental::(anonymous namespace)::ParameterizedGetNextTest_GetNext_Test>::CreateTestFactory' requested here
  TestMetaFactory() {}
  ^
third_party/tensorflow/core/kernels/data/experimental/sampling_dataset_op_test.cc:176:1: note: in instantiation of member function 'testing::internal::TestMetaFactory<tensorflow::data::experimental::(anonymous namespace)::ParameterizedGetNextTest_GetNext_Test>::TestMetaFactory' requested here
TEST_P(ParameterizedGetNextTest, GetNext) {
^
third_party/googletest/googletest/include/gtest/gtest-param-test.h:432:19: note: expanded from macro 'TEST_P'
              new ::testing::internal::TestMetaFactory<GTEST_TEST_CLASS_NAME_( \
                  ^
./third_party/tensorflow/core/kernels/data/dataset_test_base.h:488:18: note: unimplemented pure virtual method 'MakeDatasetOpKernel' in 'ParameterizedGetNextTest_GetNext_Test'
  virtual Status MakeDatasetOpKernel(
                 ^
```

frreiss committed:
Merge branch 'master' of https://github.com/tensorflow/tensorflow into issue-data-sample-test

frreiss committed:
Update test case to new format

frreiss commented:
> Fails internal tests with:
> 
> ```
> In module '//testing/base/public:gunit_headers':
> third_party/googletest/googletest/include/gtest/internal/gtest-param-util.h:401:16: error: allocating an object of abstract class type 'tensorflow::data::experimental::(anonymous namespace)::ParameterizedGetNextTest_GetNext_Test'
>     return new TestClass();
>                ^
> third_party/googletest/googletest/include/gtest/internal/gtest-param-util.h:397:12: note: in instantiation of member function 'testing::internal::ParameterizedTestFactory<tensorflow::data::experimental::(anonymous namespace)::ParameterizedGetNextTest_GetNext_Test>::CreateTest' requested here
>   explicit ParameterizedTestFactory(ParamType parameter) :
>            ^
> third_party/googletest/googletest/include/gtest/internal/gtest-param-util.h:439:16: note: in instantiation of member function 'testing::internal::ParameterizedTestFactory<tensorflow::data::experimental::(anonymous namespace)::ParameterizedGetNextTest_GetNext_Test>::ParameterizedTestFactory' requested here
>     return new ParameterizedTestFactory<TestSuite>(parameter);
>                ^
> third_party/googletest/googletest/include/gtest/internal/gtest-param-util.h:436:3: note: in instantiation of member function 'testing::internal::TestMetaFactory<tensorflow::data::experimental::(anonymous namespace)::ParameterizedGetNextTest_GetNext_Test>::CreateTestFactory' requested here
>   TestMetaFactory() {}
>   ^
> third_party/tensorflow/core/kernels/data/experimental/sampling_dataset_op_test.cc:176:1: note: in instantiation of member function 'testing::internal::TestMetaFactory<tensorflow::data::experimental::(anonymous namespace)::ParameterizedGetNextTest_GetNext_Test>::TestMetaFactory' requested here
> TEST_P(ParameterizedGetNextTest, GetNext) {
> ^
> third_party/googletest/googletest/include/gtest/gtest-param-test.h:432:19: note: expanded from macro 'TEST_P'
>               new ::testing::internal::TestMetaFactory<GTEST_TEST_CLASS_NAME_( \
>                   ^
> ./third_party/tensorflow/core/kernels/data/dataset_test_base.h:488:18: note: unimplemented pure virtual method 'MakeDatasetOpKernel' in 'ParameterizedGetNextTest_GetNext_Test'
>   virtual Status MakeDatasetOpKernel(
>                  ^
> ```

Thanks for the heads-up. It looks like this change set crossed paths with a conflicting set of changes in `dataset_test_base.h`. I have pulled the latest changes from master and updated the tests for `SamplingDataset`. I also switched over to using the new test macros. Changes are in commit 42fc1ab.
